From f06dd19ffc3e9155225e40a0eb93a73cb9e4423c Mon Sep 17 00:00:00 2001
From: Dmitry Dementyev <dementyev@google.com>
Date: Thu, 19 Dec 2024 11:02:42 -0800
Subject: [PATCH 2/3] Check account type returned by
 AbstractAccountAuthenticator.

AccountManagerService already knows which account is used during
AbstractAccountAuthenticator.getAuthToken.

KEY_ACCOUNT_NAME and KEY_ACCOUNT_TYPE in the response look unnecessary,
but we can't change API at this moment.

Bug: 364269936
Test: manual
Flag: EXEMPT bugfix
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:234a1b27f7b0af5ddd49ae52e1bdc7c07611c366)
Merged-In: Ifc62866f4feaca43abc32bc542b97f3741953f56
Change-Id: Ifc62866f4feaca43abc32bc542b97f3741953f56
---
 .../com/android/server/accounts/AccountManagerService.java  | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/services/core/java/com/android/server/accounts/AccountManagerService.java b/services/core/java/com/android/server/accounts/AccountManagerService.java
index 2ca4004c25c7..536034a5c4f2 100644
--- a/services/core/java/com/android/server/accounts/AccountManagerService.java
+++ b/services/core/java/com/android/server/accounts/AccountManagerService.java
@@ -3000,6 +3000,12 @@ public class AccountManagerService
                                         "the type and name should not be empty");
                                 return;
                             }
+                            if (!type.equals(mAccountType)) {
+                                onError(AccountManager.ERROR_CODE_INVALID_RESPONSE,
+                                        "incorrect account type");
+                                return;
+                            }
+
                             Account resultAccount = new Account(name, type);
                             if (!customTokens) {
                                 saveAuthTokenToDatabase(
-- 
2.40.1

